### PR TITLE
Update CONTRIBUTING document 

### LINKED ISSUE #414: add note about cloning core repo in contrib doc
The contributing guidelines should mention the need to clone the core repo for users to run tests locally:

```
git clone https://github.com/open-telemetry/opentelemetry-python.git opentelemetry-python-core
```

### PR DESCRIPTION
**Pull request by dhruv-vora**
Updated the CONTRIBUTING document to include steps for running tests locally.

Fixes #414

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -153,6 +153,18 @@ rather than conform to specific API names or argument patterns in the spec.
...
+## Running Tests Locally
+
+1. Go to your Contrib repo directory. `cd ~/git/opentelemetry-python-contrib`.
+2. Create a virtual env in your Contrib repo directory. `python3 -m venv my_test_venv`.
+3. Activate your virtual env. `source my_test_venv/bin/activate`.
+4. Clone the [OpenTelemetry Python](https://github.com/open-telemetry/opentelemetry-python) Python Core repo to a folder named `opentelemetry-python-core`. `git clone https://github.com/open-telemetry/opentelemetry-python.git opentelemetry-python-core`.
+5. Change directory to the repo that was just cloned. `cd opentelemetry-python-core`.
+6. Move the head of this repo to the hash you want your tests to use. This is currently the SHA `1a12fa0d681e37c1fda9cb8d46212ff3bbf6b76a` as seen in `.github/workflows/test.yml`. Use `git fetch && git checkout 1a12fa0d681e37c1fda9cb8d46212ff3bbf6b76a`.
+7. Go back to the root directory. `cd ../`.
+8. Make sure you have `tox` installed. `pip install tox`.
```
Comment (owais) [inline CONTRIBUTING.md:165]: Better to recommend installing `dev-requirements.txt` file

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -153,6 +153,18 @@ rather than conform to specific API names or argument patterns in the spec.
...
 
+## Running Tests Locally
+
+1. Go to your Contrib repo directory. `cd ~/git/opentelemetry-python-contrib`.
+2. Create a virtual env in your Contrib repo directory. `python3 -m venv my_test_venv`.
+3. Activate your virtual env. `source my_test_venv/bin/activate`.
+4. Clone the [OpenTelemetry Python](https://github.com/open-telemetry/opentelemetry-python) Python Core repo to a folder named `opentelemetry-python-core`. `git clone https://github.com/open-telemetry/opentelemetry-python.git opentelemetry-python-core`.
+5. Change directory to the repo that was just cloned. `cd opentelemetry-python-core`.
+6. Move the head of this repo to the hash you want your tests to use. This is currently the SHA `1a12fa0d681e37c1fda9cb8d46212ff3bbf6b76a` as seen in `.github/workflows/test.yml`. Use `git fetch && git checkout 1a12fa0d681e37c1fda9cb8d46212ff3bbf6b76a`.
+7. Go back to the root directory. `cd ../`.
```
Comment (owais) [inline CONTRIBUTING.md:164]: Step 5, 6 and 7 should not be mentioned here IMO. We can have a sub-section below that explains how to tests against a specific version of the core repo. Let's focus on the default case here which is to rest against master always.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -153,6 +153,18 @@ rather than conform to specific API names or argument patterns in the spec.
...
 For a deeper discussion, see: https://github.com/open-telemetry/opentelemetry-specification/issues/165
 
+## Running Tests Locally
+
+1. Go to your Contrib repo directory. `cd ~/git/opentelemetry-python-contrib`.
+2. Create a virtual env in your Contrib repo directory. `python3 -m venv my_test_venv`.
+3. Activate your virtual env. `source my_test_venv/bin/activate`.
+4. Clone the [OpenTelemetry Python](https://github.com/open-telemetry/opentelemetry-python) Python Core repo to a folder named `opentelemetry-python-core`. `git clone https://github.com/open-telemetry/opentelemetry-python.git opentelemetry-python-core`.
+5. Change directory to the repo that was just cloned. `cd opentelemetry-python-core`.
+6. Move the head of this repo to the hash you want your tests to use. This is currently the SHA `1a12fa0d681e37c1fda9cb8d46212ff3bbf6b76a` as seen in `.github/workflows/test.yml`. Use `git fetch && git checkout 1a12fa0d681e37c1fda9cb8d46212ff3bbf6b76a`.
```
Comment (owais) [inline CONTRIBUTING.md:163]: > This is currently the SHA `1a12fa0d681e37c1fda9cb8d46212ff3bbf6b76a`

This info will become stale very soon and then possibly confuse people new to the project and/or software development in general. Suggest re-wording this.

File: CONTRIBUTING.md (hunk 1 above)
Comment (dhruv-vora) [reply CONTRIBUTING.md:165]: Currently the `dev-requirements.txt` file does not have `tox` specified in it. So should `tox` be included with some specific version in the `dev-requirements.txt` file, or should it mentioned separately?

File: CONTRIBUTING.md (hunk 2 above)
Comment (dhruv-vora) [reply CONTRIBUTING.md:164]: Sure, will remove these steps

File: CONTRIBUTING.md (hunk 3 above)
Comment (dhruv-vora) [reply CONTRIBUTING.md:163]: Could you elaborate how this info is going to get updated, and how should this be reworded?

File: CONTRIBUTING.md (hunk 1 above)
Comment (owais) [reply CONTRIBUTING.md:165]: I see. That depends on whether tox internally uses dev-requirements. So if tox uses dev-requirements to install these packages in every .tox/<env>, then adding tox to this file does not make sense. In that case probably we don't need to mention creation of virtualenv above and just ask the user to install tox and leave it to the user how they want to install Python CLI tools. We should probably mention specific version of tox as well.

If tox does not use dev-dependencies and they exist to enable the "manual" use case then I think we can add tox to it.

File: CONTRIBUTING.md (hunk 3 above)
Comment (owais) [reply CONTRIBUTING.md:163]: It is updated by contributors anywhere between multiple times a day to every few days. We should reword to not mention any specific SHA or clearly state that mentioned SHA is just an example.

File: CONTRIBUTING.md (hunk 1 above)
Comment (dhruv-vora) [reply CONTRIBUTING.md:165]: According to me, tox uses dev-requirements and hence installing and running tox would be enough

**Comment by dhruv-vora** [comment]
@owais could you please review the changes

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -153,6 +153,23 @@ rather than conform to specific API names or argument patterns in the spec.
...
+## Running Tests Locally
+
+1. Go to your Contrib repo directory. `cd ~/git/opentelemetry-python-contrib`.
+2. Create a virtual env in your Contrib repo directory. `python3 -m venv my_test_venv`.
+3. Activate your virtual env. `source my_test_venv/bin/activate`.
+4. Clone the [OpenTelemetry Python](https://github.com/open-telemetry/opentelemetry-python) Python Core repo to a folder named `opentelemetry-python-core`. `git clone https://github.com/open-telemetry/opentelemetry-python.git opentelemetry-python-core`.
+5. Make sure you have `tox` installed. `pip install tox`.
+6. To sdist-package, install and test your project, just type `tox`
+
+### Run tests for a specific package
```
Comment (owais) [inline CONTRIBUTING.md:165]: Should read something like "Testing against a different Core repo branch/commit" or something else more accurate.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -153,6 +153,23 @@ rather than conform to specific API names or argument patterns in the spec.
...
 For a deeper discussion, see: https://github.com/open-telemetry/opentelemetry-specification/issues/165
 
+## Running Tests Locally
+
+1. Go to your Contrib repo directory. `cd ~/git/opentelemetry-python-contrib`.
+2. Create a virtual env in your Contrib repo directory. `python3 -m venv my_test_venv`.
+3. Activate your virtual env. `source my_test_venv/bin/activate`.
+4. Clone the [OpenTelemetry Python](https://github.com/open-telemetry/opentelemetry-python) Python Core repo to a folder named `opentelemetry-python-core`. `git clone https://github.com/open-telemetry/opentelemetry-python.git opentelemetry-python-core`.
+5. Make sure you have `tox` installed. `pip install tox`.
+6. To sdist-package, install and test your project, just type `tox`
```
Comment (owais) [inline CONTRIBUTING.md:163]: I'd replace step 6. with something like, "Run `tox` without any arguments to run tests for all packages." and then link to tox documentation.

**Comment by dhruv-vora** [comment]
@owais @lonewolf3739 could you please review the changes

**Comment by dhruv-vora** [comment]
If the changes are as required, would it be possible to merge this branch into main and close this PR?
Thank you

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -153,6 +153,23 @@ rather than conform to specific API names or argument patterns in the spec.
 
 For a deeper discussion, see: https://github.com/open-telemetry/opentelemetry-specification/issues/165
 
+## Running Tests Locally
+
+1. Go to your Contrib repo directory. `cd ~/git/opentelemetry-python-contrib`.
```
Comment (owais) [inline CONTRIBUTING.md:158]: Shouldn't assume or prescribe a location to checkout the git repo at. May be use a placeholder name, relative path or skip `cd` command entirely?

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -153,6 +153,23 @@ rather than conform to specific API names or argument patterns in the spec.
...
+2. Create a virtual env in your Contrib repo directory. `python3 -m venv my_test_venv`.
+3. Activate your virtual env. `source my_test_venv/bin/activate`.
```
Comment (owais) [inline CONTRIBUTING.md:160]: this only works for bash but may be that's good enough for the example.

Comment (owais) [reply CONTRIBUTING.md:160]: For virtualenv steps, should we link to virtualenv docs instead? Step 2. could be `create and activate virtualenv. [how to setup virtualsenv](link-to-docs)` ?

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -153,6 +153,23 @@ rather than conform to specific API names or argument patterns in the spec.
...
+4. Clone the [OpenTelemetry Python](https://github.com/open-telemetry/opentelemetry-python) Python Core repo to a folder named `opentelemetry-python-core`. `git clone https://github.com/open-telemetry/opentelemetry-python.git opentelemetry-python-core`.
+5. Make sure you have `tox` installed. `pip install tox`.
+6. Run `tox` without any arguments to run tests for all the packages. Read more about [tox](https://tox.readthedocs.io/en/latest/).
```
Comment (owais) [inline CONTRIBUTING.md:163]: If we recommend running tests with `tox`, do we even need to suggest creating a virtualenv above? In that case, we can just suggest to install and use tox. Users who don't want global tox will know how to manage it in a virtual environment.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -153,6 +153,23 @@ rather than conform to specific API names or argument patterns in the spec.
...
+
+### Testing against a different Core repo branch/commit
+
+1. Change directory to the repo that was cloned above. `cd opentelemetry-python-core`.
+2. Move the head of this repo to the SHA hash you want your tests to use. The current SHA hash can be found in `.github/workflows/test.yml`. For example, currently it is `1a12fa0d681e37c1fda9cb8d46212ff3bbf6b76a`. So use `git fetch && git checkout <current SHA hash>`.
```
Comment (owais) [inline CONTRIBUTING.md:168]: > The current SHA hash can be found in `.github/workflows/test.yml`.

This sounds like we are talking about the opentelemetry-python-core repo's .github/workflows/text.yml file which can be confusing.

**Comment by lzchen** [comment]
@dhruv-vora 
> If the changes are as required, would it be possible to merge this branch into main and close this PR?

I am not sure what you mean by this?

**Comment by lzchen** [comment]
@dhruv-vora 
Are you still working on this?

**Comment by lzchen** [comment]
@dhruv-vora 
Hey friendly ping, it would be great for us to merge this in.

**Comment by codeboten** [comment]
@owais please take a look at the updated text

**Review by ocelotl — APPROVED** [review]

**Pull request merged**